File: python/p0049_group_anagrams/solutions/array_hashtable.py
# Time complexity: O(n * k)
# Space complexity: O(n * k)
# where n = number of words, k = length of the longest word

from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        anagram_map: defaultdict = defaultdict(list)

        for word in strs:
            freq_map = [0] * 26

            for letter in word:
                freq_map[ord(letter) - ord("a")] += 1

            anagram_map[tuple(freq_map)].append(word)

        return list(anagram_map.values())

File: csharp/P0049_GroupAnagrams/Solutions/ArrayHashTable.cs
// Time complexity: O(n * k)
// Space complexity: O(n * k)
// where n = number of words, k = length of the longest word

namespace P0049_GroupAnagrams.Solutions.ArrayHashTable;

public class Solution
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        // Build anagram map
        Dictionary<String, List<String>> anagramMap = new();

        foreach (var word in strs)
        {
            // Build frequency map for each word
            int[] frequencyMap = new int[26];

            foreach (char letter in word)
                frequencyMap[letter - 'a']++;

            // Key needs to be a string
            String key = String.Join(",", frequencyMap);

            if (!anagramMap.ContainsKey(key))
                anagramMap[key] = new List<String>();

            anagramMap[key].Add(word);
        }

        return anagramMap.Values.Cast<IList<String>>().ToList();
    }
}